torchrl/envs/libs/vmas.py

```diff
@@ -0,0 +1,428 @@
+        obs, infos = self._env.reset(return_info=True)
+
+        agent_tds = []
+        for i in range(self.n_agents):
+            agent_obs = self.read_obs(obs[i])
+            agent_info = self.read_info(infos[i])
+
+            agent_td = TensorDict(
+                source={
+                    "observation": agent_obs,
+                },
+                batch_size=(self.num_envs,),
+                device=self.device,
+            )
+
+            if infos is not None:
+                agent_td.set("info", agent_info)
+            agent_tds.append(agent_td)
+
+        tensordict_out = torch.stack(agent_tds, dim=0)
+
+        return tensordict_out
+
+    def _step(
+        self,
+        tensordict: TensorDictBase,
+    ) -> TensorDictBase:
+
+        action = tensordict.get("action")
+        action = self.read_action(action)
+
+        obs, rews, dones, infos = self._env.step(action)
+
+        dones = self.read_done(dones)
+
+        agent_tds = []
+        for i in range(self.n_agents):
+            agent_obs = self.read_obs(obs[i])
+            agent_rew = self.read_reward(rews[i])
+            agent_done = dones.clone()
+            agent_info = self.read_info(infos[i])
+
+            agent_td = TensorDict(
+                source={
+                    "observation": agent_obs,
+                    "done": agent_done,
+                    "reward": agent_rew,
+                },
+                batch_size=(self.num_envs,),
+                device=self.device,
+            )
+
+            if infos is not None:
+                agent_td.set("info", agent_info)
+            agent_tds.append(agent_td)
+
+        tensordict_out = torch.stack(agent_tds, dim=0)
+
+        return tensordict_out
+
+    def read_obs(self, observations: torch.Tensor) -> torch.Tensor:
+        observations = _selective_unsqueeze(
+            observations, batch_size=torch.Size((self.num_envs,))
+        )
+        return observations
+
+    def read_info(self, infos: Dict[str, torch.Tensor]) -> torch.Tensor:
+        if len(infos) == 0:
+            return None
+        infos = TensorDict(
+            source={
+                key: _selective_unsqueeze(
+                    value, batch_size=torch.Size((self.num_envs,))
+                )
+                for key, value in infos.items()
+            },
+            batch_size=torch.Size((self.num_envs,)),
+            device=self.device,
+        )
+
+        return infos
+
+    def read_done(self, done):
+        done = _selective_unsqueeze(done, batch_size=torch.Size((self.num_envs,)))
+
+        return done
+
+    def read_reward(self, rewards):
+        rewards = _selective_unsqueeze(rewards, batch_size=torch.Size((self.num_envs,)))
+        return rewards
+
+    def read_action(self, action):
+
+        agent_actions = []
+        for i in range(self.n_agents):
+            agent_actions.append(action[i, :, ...])
+        return agent_actions
+
+    def __repr__(self) -> str:
+        return (
```

Comment:
It would be nice to test this since you implement it

Reply:
are your referring to `__repr__`?

Reply:
I assumed you were referring to __repr__ so I added a test for that. But if you were referring to smth else im happy to write more tests